cloud/services/agentpools/agentpools.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agentpools
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-02-01/containerservice"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+)
+
+// Spec contains properties to create a agent pool.
+type Spec struct {
+	Name          string
+	ResourceGroup string
+	Cluster       string
+	SKU           string
+	Replicas      int32
+	OSDiskSizeGB  int32
+}
+
+// Get fetches a agent pool from Azure.
+func (s *Service) Get(ctx context.Context, spec interface{}) (interface{}, error) {
+	agentPoolSpec, ok := spec.(*Spec)
+	if !ok {
+		return containerservice.AgentPool{}, errors.New("expected agent pool specification")
+	}
+	return s.Client.Get(ctx, agentPoolSpec.ResourceGroup, agentPoolSpec.Cluster, agentPoolSpec.Name)
+}
+
+// Reconcile idempotently creates or updates a agent pool, if possible.
+func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
+	agentPoolSpec, ok := spec.(*Spec)
+	if !ok {
+		return errors.New("expected agent pool specification")
+	}
+
+	profile := containerservice.AgentPool{
+		ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
+			VMSize:       containerservice.VMSizeTypes(agentPoolSpec.SKU),
+			OsDiskSizeGB: &agentPoolSpec.OSDiskSizeGB,
+			Count:        &agentPoolSpec.Replicas,
+			Type:         containerservice.VirtualMachineScaleSets,
+		},
+	}
+
+	existingSpec, err := s.Get(ctx, spec)
+	existingPool, ok := existingSpec.(containerservice.AgentPool)
+	if !ok {
+		return errors.New("expected agent pool specification")
+	}
+
+	if err == nil {
+		existingProfile := containerservice.AgentPool{
+			ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
+				VMSize:       existingPool.ManagedClusterAgentPoolProfileProperties.VMSize,
+				OsDiskSizeGB: existingPool.ManagedClusterAgentPoolProfileProperties.OsDiskSizeGB,
+				Count:        existingPool.ManagedClusterAgentPoolProfileProperties.Count,
+				Type:         containerservice.VirtualMachineScaleSets,
+			},
+		}
+
+		diff := cmp.Diff(profile, existingProfile)
+		if diff != "" {
+			fmt.Printf("update required (+new -old):\n%s", diff)
```

Comment:
Should this be a log statement?

Reply:
I was actually torn on this -- the whole point of this diff is pretty printing the difference, which structured logging will totally swallow..but I guess we shouldn't be encouraging people to manually pull out unstructured logs? This is maybe more appropriate for tests

Reply:
second opinion on this one?

Reply:
+1 on log

It should probably be .V(2+) too